client_int_test.go

```diff
@@ -0,0 +1,289 @@
+// +build integration
+
+package function_test
+
+import (
+	"os"
+	"reflect"
+	"testing"
+	"time"
+
+	boson "github.com/boson-project/func"
+	"github.com/boson-project/func/buildpacks"
+	"github.com/boson-project/func/docker"
+	"github.com/boson-project/func/knative"
+)
+
+/*
+ NOTE:  Running integration tests locally requires a configured test cluster.
+        Test failures may require manual removal of dangling resources.
+
+ ## Integration Cluster
+
+ These integration tests require a properly configured cluster,
+ such as that which is setup and configured in CI (see .github/workflows).
+ A local KinD cluster can be started via:
+   ./hack/allocate.sh && ./hack/configure.sh
+
+ ## Integration Testing
+
+ These tests can be run via the make target:
+   make test-integration
+  or manually by specifying the tag
+   go test -v -tags integration ./...
+
+ ## Teardown and Cleanup
+
+ Tests should clean up after themselves.  In the event of failures, one may
+ need to manually remove files:
+   rm -rf ./testdata/example.com
+ The test cluster is not automatically removed, as it can be reused.  To remove:
+   ./hack/delete.sh
+*/
+
+const (
+	// DefaultRegistry must contain both the registry host and
+	// registry namespace at this time.  This will likely be
+	// split and defaulted to the forthcoming in-cluster registry.
+	DefaultRegistry = "localhost:5000/func"
+
+	// DefaultNamespace for the underlying deployments.  Must be the same
+	// as is set up and configured (see hack/configure.sh)
+	DefaultNamespace = "func"
+)
+
+func TestList(t *testing.T) {
+	verbose := true
+
+	// Assemble
+	lister, err := knative.NewLister(DefaultNamespace)
+	if err != nil {
+		t.Fatal(err)
+	}
+	client := boson.New(
+		boson.WithLister(lister),
+		boson.WithVerbose(verbose))
+
+	// Act
+	names, err := client.List()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Assert
+	if len(names) != 0 {
+		t.Fatalf("Expected no Functions, got %v", names)
+	}
+}
+
+// TestNew creates
+func TestNew(t *testing.T) {
+	defer within(t, "testdata/example.com/testnew")()
+	verbose := true
+
+	client := newClient(verbose)
+
+	// Act
+	if err := client.New(boson.Function{Name: "testnew", Root: ".", Runtime: "go"}); err != nil {
+		t.Fatal(err)
+	}
+	defer del(t, client, "testnew")
+
+	// Assert
+	items, err := client.List()
+	names := []string{}
+	for _, item := range items {
+		names = append(names, item.Name)
+	}
+	if err != nil {
+		t.Fatal(err)
+	}
+	if !reflect.DeepEqual(names, []string{"testnew"}) {
+		t.Fatalf("Expected function list ['testnew'], got %v", names)
+	}
+}
+
+// TestDeploy updates
+func TestDeploy(t *testing.T) {
+	defer within(t, "testdata/example.com/deploy")()
+	verbose := true
+
+	client := newClient(verbose)
+
+	if err := client.New(boson.Function{Name: "deploy", Root: ".", Runtime: "go"}); err != nil {
+		t.Fatal(err)
+	}
+	defer del(t, client, "deploy")
+
+	if err := client.Deploy("."); err != nil {
+		t.Fatal(err)
+	}
```

Comment:
I'd expect this test to hit the deployed app and ensure that it's served, do you plan to do that?

Reply:
Sure am; thanks for noticing.  That's the part I am finishing up before pulling this PR out of draft status.